k8s-plugin-keel/src/main/kotlin/com/amazon/spinnaker/keel/k8s/verificationEvaluator/K8sJobEvaluator.kt

```diff
@@ -0,0 +1,182 @@
+                }
+                response.status.isIncomplete() -> oldState.copy(ConstraintStatus.PENDING).also {
+                    log.debug("verification still in progress. taskId: $taskId")
+                }
+                else -> oldState.copy(ConstraintStatus.FAIL).also {
+                    log.warn("verification failed. taskId: $taskId")
+                }
+            }
+        } ?: return oldState.copy(status = ConstraintStatus.FAIL)
+    }
+
+    override fun start(context: ArtifactInEnvironmentContext, verification: Verification): Map<String, Any?> {
+        require(verification is K8sJobVerification) {
+            "verification class must be ${K8sJobVerification::class.simpleName}. received: ${verification.javaClass.simpleName}"
+        }
+        log.info("launching verification K8s job for ${context.deliveryConfig.application} env: ${context.environmentName}")
+        val (job, metadata) = generateOrcaK8sJob(
+            context.deliveryConfig.application,
+            context.environmentName,
+            verification
+        )
+
+        log.info("this job will be launched in namespace, ${metadata[NAMESPACE]}. Name: ${metadata["jobName"]}")
+        return runBlocking {
+            val launcherResponse = taskLauncher.submitJob(
+                user = context.deliveryConfig.serviceAccount,
+                application = context.deliveryConfig.application,
+                notifications = emptySet(),
+                subject = "verify environment, ${context.environmentName}, for ${context.deliveryConfig.application}",
+                description = "verifying ${context.version} in environment ${context.environmentName} with K8s $VERIFICATION_K8S_JOB_V1",
+                correlationId = verification.id,
+                stages = listOf(job),
+                type = VERIFICATION,
+                artifacts = emptyList(),
+                parameters = emptyMap()
+            )
+            metadata += ("taskId" to launcherResponse.id)
+            metadata + ("taskName" to launcherResponse.name)
+        }
+    }
+
+    // need to randomize name otherwise subsequent jobs may fail with the same name in the same namespace
+    // neither orca, clouddriver, nor keel cleans finished jobs.
+    private fun generateOrcaK8sJob(
+        application: String,
+        environment: String,
+        verification: K8sJobVerification
+    ): Pair<OrcaJob, MutableMap<String, String>> {
+        verification.manifest[API_VERSION] = VERIFICATION_K8S_JOB_API_V1
+        verification.manifest[KIND] = VERIFICATION_K8S_JOB_KIND
+
+        val metadata = if (verification.manifest.containsKey("metadata")) {
+            @Suppress("UNCHECKED_CAST")
+            verification.manifest["metadata"] as MutableMap<String, Any>
+        } else {
+            mutableMapOf()
+        }
+
+        metadata[NAME] = generateJobName(application, environment, verification.jobNamePrefix)
```

Comment:
you can probably make this whole lot simpler and remove bunch of code by using `generateName` for your Kubernetes jobs:

```
apiVersion: batch/v1
kind: Job
metadata:
  generateName: kube-job-
  labels:
    jobgroup: jobexample
```

more details here: https://kubernetes.io/docs/reference/using-api/api-concepts/#generated-values

Reply:
I thought about it but we don't have the job name until it's submitted and would be difficult to track which kubernetes job belongs to which verifyWith job. The metadata generated as a result of `start()` stores this information and can be used later for troubleshooting purposes.

Reply:
hmm interesting. Do we have the job name from when we get the task response back from Orca? if so, we might be able to correlate.

Also how important is it for us to be able to track individual job runs back to their executions. The prefix will identify the `verifyWith` task uniquely. Do we care about differentiatings across multiple runs of the same `verifyWith`? can we make better use of timestamps etc?

Reply:
task ID and task name are already stored when we submit the job. We can correlate verifyWith with orca task but that's about it. The prefix probably should not be relied on to identify tasks uniquely. Users will copy/paste configurations. 
timestamp could work but it's possible for them to collide when executed at the same time. 
Same verifyWith will not run at the same time but multiple verifyWith tasks per application per environment can ran at the same time. 
If we don't care about tracing verityWith execution to k8s job, we can use `generateName`. Maybe we can just stick some diagnostic information into Jobs' metadata like execution time, env name, and artifact version. 
Thinking about it, this might be a better approach.

Reply:
yeah I like where you are going with adding metadata to k8s jobs to identify tasks.